File: rsi/evidence/2026-09-21/capstone-portability/driver.mjs
// Two author smoke runs through a fixed shared tool and canonical skill.
import {readFileSync,writeFileSync,mkdirSync,existsSync,copyFileSync,readdirSync} from 'node:fs';
import {resolve,dirname,relative} from 'node:path';
import {fileURLToPath} from 'node:url';
import {createHash} from 'node:crypto';
import {spawnSync} from 'node:child_process';
const repo=resolve(dirname(fileURLToPath(import.meta.url)),'../../..');
const root=resolve(repo,'../rsi-work-2026-09-21-capstone-portability');
const python=resolve(repo,'.venv/Scripts/python.exe'),lab=resolve(repo,'rsi/tools/lab.py'),checker=resolve(repo,'rsi/tools/check_result.py');
const hash=p=>createHash('sha256').update(readFileSync(p)).digest('hex');
const csv=(fields,rows)=>fields.join(',')+'\n'+rows.map(r=>fields.map(k=>'"'+String(r[k]).replaceAll('"','""')+'"').join(',')).join('\n')+'\n';
const put=(p,s)=>{if(existsSync(p))throw Error('Refuse overwrite '+p);writeFileSync(p,s);};
const paths=base=>readdirSync(base,{withFileTypes:true}).flatMap(e=>e.name==='__pycache__'?[]:e.isDirectory()?paths(resolve(base,e.name)):[resolve(base,e.name)]);
function run(label,args,expected=0){
  const path=resolve(root,'commands',label+'.md');if(existsSync(path))throw Error('Command already recorded');
  const started=new Date().toISOString(),t=performance.now();
  const r=spawnSync(python,args,{cwd:repo,encoding:'utf8',timeout:60000,windowsHide:true});
  put(path,`# ${label}\n\nStarted: ${started}\n\nExecutable: ${python}\n\nArguments: ${args.map(x=>'`'+x+'`').join(' ')}\n\nExit: ${r.status}; signal: ${r.signal}; seconds: ${(performance.now()-t)/1000}\n\n\`\`\`text\n${r.stdout||''}${r.stderr||''}${r.error?String(r.error):''}\n\`\`\`\n`);
  console.log(label+': '+r.status+' '+(r.stdout||r.stderr||'').trim());
  if(r.status!==expected)throw Error('Unexpected status '+label);
}
switch(process.argv[2]){
  case 'prepare':{
    if(existsSync(root))throw Error('Workspace exists');mkdirSync(resolve(root,'commands'),{recursive:true});mkdirSync(resolve(root,'sources'));
    copyFileSync(fileURLToPath(import.meta.url),resolve(root,'driver.mjs'));
    const inputs=[['PROTOCOL.md','how-did-i-generate-it/rsi/validation/CAPSTONE-PORTABILITY-PROTOCOL.md'],
      ['RUN-SKILL.md','rsi/skills/run-ml-experiment/SKILL.md'],['TUTOR-SKILL.md','rsi/skills/rsi-tutor/SKILL.md'],
      ['BIKE-DATA-CARD.md','rsi/examples/bike-demand/DATA-CARD.md'],['WINE-DATA-CARD.md','rsi/examples/wine-quality/DATA-CARD.md'],
      ['sources/lab.py','rsi/tools/lab.py'],['sources/check_result.py','rsi/tools/check_result.py'],['sources/requirements.txt','rsi/tools/requirements.txt'],
      ['sources/hour.csv','rsi/examples/bike-demand/source/hour.csv'],['sources/winequality-red.csv','rsi/examples/wine-quality/source/winequality-red.csv']];
    const records=[];
    for(const [destination,source] of inputs){copyFileSync(resolve(repo,source),resolve(root,destination));records.push({source,destination,sha256:hash(resolve(repo,source))});}
    put(resolve(root,'SOURCES.csv'),csv(['source','destination','sha256'],records));
    put(resolve(root,'DECISION.md'),'# Before the two fits\n\nThe author read the canonical skill and both data cards. Use explicit task, all permitted features, linear model, seed 17 and one attempt per task. The same Markdown procedure governs data inspection, bounded execution and checking; source snapshots alone do not prove compliance.\n\nBike hypothesis: calendar and observed weather can support a simple retrospective rental-count estimate. This is not tomorrow\'s forecast. Metric: MAE, smaller is better; chronological split.\n\nWine hypothesis: a balanced logistic classifier on permitted chemistry inputs provides an executable baseline for the quality-at-least-seven label. Metric: balanced accuracy, larger is better; feature-group split; inspect both recalls.\n\nThese are smoke hypotheses, not comparative gain claims. Model families differ because targets differ. No I1 transfer or new autonomous research is tested. Learner predictions and quiz responses are unattempted.\n');
    run('00-environment',['-c',"import sys,platform,importlib.metadata as m; print(sys.version); print(platform.platform()); print(*[k+'='+m.version(k) for k in ['numpy','pandas','scikit-learn','scipy','matplotlib']],sep=chr(10))"]);
    break;
  }
  case 'execute':{
    const invalid=resolve(root,'unsupported-task');
    run('01-invalid-task',[lab,'run','--task','unsupported','--workspace',invalid,'--hypothesis','Refusal before allocation'],2);
    if(existsSync(invalid))throw Error('Invalid task created a workspace');
    put(resolve(root,'RECOVERY.md'),'# Invalid request recovered\n\nThe unsupported task name was refused by argparse before any workspace or attempt was created. Correcting the name selects the declared bike task. It does not refund a charged failure; none was admitted. Continue the original one-fit-per-task allocation.\n');
    for(const task of ['bike','wine']){
      const work=resolve(root,task),policy=resolve(root,'RUN-SKILL.md');
      run(task+'-inspect',[lab,'inspect','--task',task,'--workspace',work]);
      const args=[lab,'run','--task',task,'--workspace',work,'--model','linear','--features','all','--seed','17','--attempt-limit','1','--policy',policy,'--hypothesis',task==='bike'?'Calendar and observed weather support a simple retrospective demand estimate':'Balanced logistic classification supplies a chemistry-based quality-threshold smoke baseline'];
      run(task+'-fit',args);
      run(task+'-check',[checker,resolve(work,'trial-001'),'--report',resolve(work,'trial-001/CHECK.md')]);
      run(task+'-resume-compare',[lab,'compare','--workspace',work]);
      run(task+'-extra-fit-refused',args,2);
      if(task==='wine'){
        const changed=[...args];changed[changed.indexOf('--attempt-limit')+1]='2';run('wine-budget-change-refused',changed,2);
      }
    }
    break;
  }
  case 'seal':{
    const files=paths(root);put(resolve(root,'MANIFEST.csv'),csv(['path','sha256'],files.map(p=>({path:relative(root,p).replaceAll('\\','/'),sha256:hash(p)}))));console.log('Sealed '+files.length+' files.');break;
  }
  default:throw Error('Expected prepare, execute or seal');
}

File: how-did-i-generate-it/rsi/validation/fresh-session-results/complete/driver.py
"""One-attempt maintenance driver. Invoke preflight, then attempt, never retry."""
import hashlib
import importlib.metadata
import json
import os
from pathlib import Path
import platform
import subprocess
import sys
import time
from datetime import datetime, timezone

OUT = Path(__file__).resolve().parent
REPO = OUT.parent / 'simple-coding-harness'
PYTHON = REPO / '.venv/Scripts/python.exe'
PACKET = REPO / 'how-did-i-generate-it/rsi/validation/fresh-session-plan/complete'
LOG = OUT / 'commands'
LOG.mkdir(exist_ok=True)

def write(name, value):
    (OUT / name).write_text(value, encoding='utf-8')

def command(name, args):
    folder = LOG / name
    folder.mkdir(exist_ok=False)
    meta = dict(argv=[str(x) for x in args], cwd=str(REPO), timeout_seconds=60,
                started_utc=datetime.now(timezone.utc).isoformat())
    (folder / 'command.json').write_text(json.dumps(meta, indent=2), encoding='utf-8')
    start = time.monotonic()
    env = dict(os.environ, PYTHONDONTWRITEBYTECODE='1')
    with (folder / 'stdout.txt').open('wb') as stdout, (folder / 'stderr.txt').open('wb') as stderr:
        child = subprocess.Popen(args, cwd=REPO, stdout=stdout, stderr=stderr, env=env)
        meta['pid'] = child.pid
        try:
            meta['exit_code'] = child.wait(timeout=60)
            meta['timed_out'] = False
        except subprocess.TimeoutExpired:
            child.kill()
            meta['exit_code'] = child.wait()
            meta['timed_out'] = True
    meta['elapsed_seconds'] = time.monotonic() - start
    meta['ended_utc'] = datetime.now(timezone.utc).isoformat()
    (folder / 'exit.json').write_text(json.dumps(meta, indent=2), encoding='utf-8')
    print(json.dumps(meta), flush=True)
    return meta['exit_code'] == 0 and not meta['timed_out']

def identities(name):
    paths = [PACKET / x for x in ('HANDOFF.md', 'BASELINE-SKILL.md', 'TASK.md')]
    paths += [REPO / x for x in ('rsi/AGENTS.md', 'rsi/tools/README.md', 'rsi/tools/requirements.txt',
        'rsi/tools/lab.py', 'rsi/tools/check_result.py', 'rsi/examples/bike-demand/DATA-CARD.md',
        'rsi/examples/bike-demand/source/hour.csv', 'rsi/examples/bike-demand/source/bike-sharing.zip')]
    paths += [PYTHON, Path(__file__).resolve()]
    hashes = {str(p): hashlib.sha256(p.read_bytes()).hexdigest() for p in paths}
    write(name, json.dumps(hashes, indent=2))
    assert hashes[str(PACKET / 'BASELINE-SKILL.md')] == '210bed6c0528855a878771f8c8c9acd635a7af1da38ab6f8d15a9ab1946e1d2b'
    assert hashes[str(REPO / 'rsi/examples/bike-demand/source/hour.csv')] == 'e03de4ee4ef4dc376ac6e04bf829673c6269e8eba5c60fa121640fa2f829504f'
    assert hashes[str(REPO / 'rsi/examples/bike-demand/source/bike-sharing.zip')] == 'b70182d0d0508e9abbb79306ce5c0cec34869000f8220175ac83d11dbe845401'

def main():
    assert Path(sys.executable).resolve() == PYTHON.resolve(), sys.executable
    mode = sys.argv[1]
    if mode == 'preflight':
        (OUT / 'PREFLIGHT-STARTED').open('x').close()
        identities('identities-before.json')
        versions = {'python': sys.version, 'executable': sys.executable, 'platform': platform.platform(),
                    'packages': {d.metadata['Name']: d.version for d in importlib.metadata.distributions()}}
        write('dependencies.json', json.dumps(versions, indent=2, sort_keys=True))
        assert command('01-commit', ['git', 'rev-parse', 'HEAD'])
        commit = (LOG / '01-commit/stdout.txt').read_text().strip()
        assert commit == '0132b2302e3e0212fb233f229322a09e8f2c2a88'
        assert command('02-inspect', [PYTHON, REPO / 'rsi/tools/lab.py', 'inspect', '--task', 'bike', '--workspace', OUT])
        write('PROGRESS.md', '# Progress\n\nPreflight and data inspection completed. No baseline attempt yet. One attempt remains.\n')
    elif mode == 'attempt':
        (OUT / 'BASELINE-ATTEMPT-STARTED').open('x').close()
        ok = command('03-baseline', [PYTHON, REPO / 'rsi/tools/lab.py', 'run', '--task', 'bike',
            '--workspace', OUT, '--model', 'constant', '--features', 'calendar', '--seed', '17',
            '--attempt-limit', '1', '--hypothesis', 'Use the training median as the fixed hourly-rental baseline.',
            '--policy', PACKET / 'BASELINE-SKILL.md'])
        checked = False
        if ok:
            checked = command('04-check', [PYTHON, REPO / 'rsi/tools/check_result.py', OUT / 'trial-001',
                              '--report', OUT / 'trial-001/CHECK.md'])
        identities('identities-after.json')
        write('PROGRESS.md', '# Progress\n\nThe sole baseline attempt has ended. Baseline success: ' + str(ok) +
              '. Course prediction check passed: ' + str(checked) + '. No attempts remain. No final evaluation was run.\n')
        return 0 if ok and checked else 1
    else:
        raise ValueError('Only preflight or attempt is allowed')
    return 0

if __name__ == '__main__':
    raise SystemExit(main())
